site/content/blog/crossplane-ytt-integration.md

```diff
@@ -0,0 +1,147 @@
+---
+title: "Using ytt to create Crossplane Template Function"
+slug: crossplane-ytt-template-function
+date: 2024-02-05
+author: Rohit Aggarwal
+excerpt: "Using ytt to create composition functions to template Crossplane resources"
+image: /img/logo.svg
+tags: ['Rohit Aggarwal', 'ytt', 'crossplane']
+---
+
+## What is Crossplane?
+
+
+[Crossplane](https://www.crossplane.io/) is an open-source Kubernetes extension that empowers organizations to manage cloud infrastructure across any cloud through standard Kubernetes APIs. It allows platform teams to declaratively define and manage the cloud infrastructure, like  databases, storage volumes, virtual machines, etc., through Kubernetes APIs. 
+
+## Crossplane Composite Resource
+
+A [composite resource](https://docs.crossplane.io/v1.14/concepts/composite-resources/) represents a set of managed resources as a single Kubernetes object. It uses Composition template to create multiple managed resources as a single Kubernetes object. 
+
+## Crossplane Compositions
+
+[Compositions](https://docs.crossplane.io/v1.14/concepts/compositions/) are a template for creating multiple managed resources as a single object. A Composition composes individual managed resources together into a larger, reusable solution. An example of Composition may combine a virtual machine, storage resources, and networking policies. A Composition template links all these individual resources together.
+
+Crossplane Compositions have some limitations though:
+
+1. Compositions don't support conditions, meaning that the transformations they provide are applied on an "all or nothing" basis.
+2. They also don't support loops, which means that you cannot apply transformations iteratively.
+3. Finally, advanced operations are not supported either, like checking for statuses in other systems.
+
+To overcome this, crossplane introduced the composition function.
+
+## Crossplane Composition function
+
+[Composition functions](https://docs.crossplane.io/v1.14/concepts/composition-functions/) (or just functions, for short) are custom programs that template Crossplane resources. Crossplane calls composition functions to determine what resources it should create when you create a composite resource (XR).
+
+## Prerequisites
+* kind -- to create a test Kubernetes cluster. Use this [link](https://kind.sigs.k8s.io/docs/user/quick-start/#creating-a-cluster) to set up Kubernetes cluster on `kind`.
+* kubectl -- to apply manifests to provision cloud resources;
+* Helm -- to install Crossplane on the test cluster.
+
+## Install Crossplane
+Use this [link](https://docs.crossplane.io/latest/software/install/) to install crossplane.
+
+
+## Create the ytt templating composition function
+
+[ytt](https://carvel.dev/ytt) is a tool to template and patch YAML files written in [Go](https://go.dev/). Mostly, it has been used as an executable in the form of cli. However, in this example we will use it as a library to create crossplane composition function.
```

Comment:
```suggestion
[Ytt](https://carvel.dev/ytt) is a tool to template and patch YAML files written in [Go](https://go.dev/). Mostly, it has been used as an executable in the form of CLI. However, in this example we will use it as a library to create Crossplane Composition function.
```

Reply:
ytt is capitalized when it's the beginning of a sentence

Reply:
Resolved